fdbclient/ManagementAPI.actor.cpp

```diff
@@ -2115,34 +2117,79 @@ ACTOR Future<Void> lockDatabase(Transaction* tr, UID id) {
 		}
 	}
 
-	tr->atomicOp(databaseLockedKey,
+	tr->atomicOp(lockKey,
 	             BinaryWriter::toValue(id, Unversioned()).withPrefix("0123456789"_sr).withSuffix("\x00\x00\x00\x00"_sr),
 	             MutationRef::SetVersionstampedValue);
-	tr->addWriteConflictRange(normalKeys);
+	tr->addWriteConflictRange(affectedRange);
 	return Void();
 }
 
-ACTOR Future<Void> lockDatabase(Reference<ReadYourWritesTransaction> tr, UID id) {
+ACTOR template <class Tr>
+Future<Void> unlockDBOrTenant(Tr* tr, UID id, StringRef lockKey) {
 	tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
 	tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-	Optional<Value> val = wait(tr->get(databaseLockedKey));
+	Optional<Value> val = wait(tr->get(lockKey));
 
-	if (val.present()) {
-		if (BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()) == id) {
-			return Void();
-		} else {
-			//TraceEvent("DBA_LockLocked").detail("Expecting", id).detail("Lock", BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()));
-			throw database_locked();
-		}
+	if (!val.present())
+		return Void();
+
+	if (val.present() && BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()) != id) {
+		//TraceEvent("DBA_UnlockLocked").detail("Expecting", id).detail("Lock", BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()));
+		throw database_locked();
 	}
 
-	tr->atomicOp(databaseLockedKey,
-	             BinaryWriter::toValue(id, Unversioned()).withPrefix("0123456789"_sr).withSuffix("\x00\x00\x00\x00"_sr),
-	             MutationRef::SetVersionstampedValue);
-	tr->addWriteConflictRange(normalKeys);
+	tr->clear(singleKeyRange(lockKey));
 	return Void();
 }
 
+template <class Tr>
+Future<Void> lockTenantImpl(Tr* tr, UID id, int64_t tenantID, bool rawAccess) {
+	Arena arena;
+	StringRef lockKey = databaseLockedKey;
+	KeyRangeRef affectedRange = normalKeys;
+	if (rawAccess) {
+		StringRef tenantPrefix(reinterpret_cast<const uint8_t*>(tenantID), sizeof(int64_t));
+		lockKey = databaseLockedKey.withPrefix(tenantPrefix, arena);
+		affectedRange = KeyRangeRef(tenantPrefix, tenantPrefix.withSuffix("\xff"_sr, arena));
```

Comment:
Can we use prefix range for this? Or if not, I think we should include through the `\xff\xff` suffix. We can't write to those keys now, but it will be more future-proof that way.